SignalServiceKit/src/Messages/OWSMessageManager.m

```diff
@@ -1162,7 +1095,36 @@ - (void)finalizeIncomingMessage:(TSIncomingMessage *)incomingMessage
         [incomingMessage markAsReadWithTransaction:transaction sendReadReceipt:NO updateExpiration:YES];
     }
 
-    DDLogDebug(@"%@ shouldMarkMessageAsRead: %d (%@)", self.logTag, shouldMarkMessageAsRead, envelope.source);
+    TSQuotedMessage *_Nullable quotedMessage = incomingMessage.quotedMessage;
+    if (quotedMessage && quotedMessage.thumbnailAttachmentPointerId) {
+        // We weren't able to derive a local thumbnail, so we'll fetch the referenced attachment.
+        TSAttachmentPointer *attachmentPointer =
+            [TSAttachmentPointer fetchObjectWithUniqueID:quotedMessage.thumbnailAttachmentPointerId
+                                             transaction:transaction];
+
+        if ([attachmentPointer isKindOfClass:[TSAttachmentPointer class]]) {
+            OWSAttachmentsProcessor *attachmentProcessor =
+                [[OWSAttachmentsProcessor alloc] initWithAttachmentPointer:attachmentPointer
+                                                            networkManager:self.networkManager];
+
+            DDLogDebug(@"%@ downloading thumbnail for message: %tu", self.logTag, incomingMessage.timestamp);
+            [attachmentProcessor fetchAttachmentsForMessage:incomingMessage
+                transaction:transaction
+                success:^(TSAttachmentStream *_Nonnull attachmentStream) {
+                    [self.dbConnection
+                        asyncReadWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
+                            [incomingMessage setQuotedMessageThumbnailAttachmentStream:attachmentStream];
+                            [incomingMessage saveWithTransaction:transaction];
+                        }];
+                }
+                failure:^(NSError *_Nonnull error) {
+                    DDLogWarn(@"%@ failed to fetch thumbnail for message: %tu with error: %@",
```

Comment:
We don't currently handle retry. 

Keep in mind, most quotes will be a recent message, which allows the thumbnail to be generated locally rather than fetched over the network, so I think it makes sense to prefer a simple rather than perfect solution here.

With attachments the pattern is for a tap-to-retry. We could do something similar with thumbnail attachments. WDYT @charlesmchen?